File: ahps_web/static/js/modules.js
/*
#
# AHPS Web - web server for managing an AtHomePowerlineServer instance
# Copyright (C) 2014, 2015  Dave Hocker (email: AtHomeX10@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the LICENSE file for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program (the LICENSE file).  If not, see <http://www.gnu.org/licenses/>.
#
*/

/*
    Home page app controller
*/
app.controller('modulesController', function($scope, $http, Tracker) {
    // Initialization

    $scope.title = "AHPS Web";
    $scope.error = "";
    $scope.message = "";
    $scope.roomid = $("#roomid").val();
    $scope.room_modules = [];

    $scope.track_change = Tracker.track_change;

    /* Get all modules for the current house */
    $scope.get_modules = function() {
        $http.get('/modules/' + String($scope.roomid), {}).
            success(function(data, status, headers, config) {
                $scope.room_modules = data.modules;
            }).
            error(function(data, status, headers, config) {
                $scope.status = "";
                $scope.error = data.message;
            });
    };

    $scope.is_module_type_match = function(module_type, mt) {
        return mt.toLowerCase() == module_type.toLowerCase();
    };

    $scope.is_house_code_match = function(module_hc, hc) {
        return hc.toLowerCase() == module_hc.toLowerCase();
    };

    /* When the module type changes, immediately save the module
    so the type change is obvious */
    $scope.module_type_changed = function(module) {
        console.log("Module type was changed");
        $scope.save_module(module);
    }

    $scope.house_code_changed = function(moduleid) {
    };

    $scope.device_code_changed = function(moduleid) {
    };

    /* Save all module properties */
    $scope.save_module = function(module) {
        var moduleid = module.moduleid;
        var rp = {};
        rp["module_type"] = module["module_type"].toLowerCase();
        rp["name"] = module["name"];
        rp["house_code"] = module["house_code"];
        rp["device_code"] = module["device_code"];
        rp["dim_amount"] = module["dim_amount"];

        $http.put('/module/' + String(module.moduleid), {"data": rp}).
            success(function(data, status, headers, config) {
                // Success
                $scope.error = "";
                $scope.message = "Saved";
                window.onbeforeunload = null;
            }).
            error(function(data, status, headers, config) {
                if (data && data.message) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Error saving module"
                }
            });
    };

    $scope.show_programs = function(moduleid) {
        window.location.replace('/module/' + String(moduleid) + '/programs/page');
    };

    $scope.module_on = function(moduleid) {
        $http.put('/modules/' + String(moduleid) + '/state' , {"data": {"state": "on"}}).
            success(function(data, status, headers, config) {
                // Success
                $scope.error = "";
            }).
            error(function(data, status, headers, config) {
                if (data && data.message) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Error turning on module"
                }
            });
    };

    $scope.module_off = function(moduleid) {
        $http.put('/modules/' + String(moduleid) + '/state' , {"data": {"state": "off"}}).
            success(function(data, status, headers, config) {
                // Success
                $scope.error = "";
            }).
            error(function(data, status, headers, config) {
                if (data && data.message) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Error turning off module"
                }
            });
    };

    $scope.show_remove_dialog = function(moduleid) {
        // Show confirmation dialog for removing a module #}
        $("#dialog").data("moduleid", moduleid).dialog("open");
    };

    $scope.angular_delete = function() {
        // Refresh page
        $scope.get_modules();
    };

    $scope.show_move_dialog = function(moduleid) {
        // Show  dialog for moving a module
        $("#move-module-dialog").data("moduleid", moduleid).dialog("open");
    };

    // Move a module to another room
    $scope.angular_move = function(moduleid) {
        var to_room = $("#new_room").val();
        $http.post('/rooms/' + String(to_room) + "/module/" + String(moduleid), {}).
            success(function(data, status, headers, config) {
                // Success
                $scope.error = "";
                // After move, refresh page
                $scope.get_modules();
            }).
            error(function(data, status, headers, config) {
                if (data && data.message) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Error moving module"
                }
            });
    };

    $scope.add_appliance = function(roomid) {
        window.location.replace("/room/" + String(roomid) + "/new_appliance_module");
    };

    $scope.add_lamp = function(roomid) {
        window.location.replace("/room/" + String(roomid) + "/new_lamp_module");
    };

    /* More initialization */
    $scope.get_modules();

});

$(document).ready(function() {
    /* Initialize the confirmation dialog */
    $("#dialog").dialog(
    {
        autoOpen: false,
        modal: true,
        closeOnEscape: false,
        width: 400,
        buttons: {
            "Remove": function(event) {
                $("#dialog").dialog( "close" );
                // $("#module-form-" + $(this).data("moduleid")).submit();

                $.ajax({
                    url: '/module/' + $(this).data("moduleid"),
                    type: 'DELETE',
                    success: function() {
                        console.log("Successful module delete");
                        // Get the scope for the html element
                        // and call the delete event.
                        // This is a bit of a hack, but to get rid
                        // of it will require replacing all of the dialogs.
                        var el = $("html");
                        var scope = angular.element(el).scope();
                        scope.angular_delete();
                    }
                });
            },
            "Cancel": function(event) {
                $("#dialog").dialog( "close" );
            }
        }
    });

    /* Initialize the module-move dialog box */
    $("#move-module-dialog").dialog(
    {
        autoOpen: false,
        modal: true,
        closeOnEscape: false,
        buttons: {
            "Move": function(event) {
                $("#move-module-dialog").dialog( "close" );
                var el = $("html");
                var scope = angular.element(el).scope();
                scope.angular_move($(this).data("moduleid"));
            },
            "Cancel": function(event) {
                $("#move-module-dialog").dialog( "close" );
            }
        }
    });

});

File: ahps_web/views/modules.py
#
# AHPS Web - web server for managing an AtHomePowerlineServer instance
# Copyright (C) 2014, 2015  Dave Hocker (email: AtHomeX10@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the LICENSE file for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program (the LICENSE file).  If not, see <http://www.gnu.org/licenses/>.
#
from ahps_web import app
from datetime import datetime
from flask import Flask, request, session, g, redirect, url_for, abort, \
    render_template, flash, jsonify
from ahps_web.models.room import get_rooms, get_room, insert_room, delete_room, update_room
from ahps_web.models.module import get_modules_for_room, get_module, insert_module, update_module_hdc, \
    update_module_dim_amount, update_module_name, delete_module, delete_room_modules, get_modules_for_house, \
    update_module_type, move_module_room, update_module_selected
from ahps_web.models.program import get_program, get_programs_for_module, insert_program, delete_program, \
    update_program, delete_module_programs
from ahps_web.models.house_programs import get_house_summary
from ahps_web.bll.download import Downloader
from ahps_web.bll.sun_data import get_sun_data
from ahps_web.models.house import get_current_house, get_houses, set_current_house, get_house, update_house, \
    insert_house, delete_house
from ahps_web.bll.x10_control import device_on, device_off, all_lights_off, all_lights_on
from ahps_web.bll.copy_house import copy_house as bll_copy_house
from flask_login import login_required
# import view_helpers # register context processors
import json


@app.route('/modules_page/<roomid>', methods=['GET'])
@login_required                                 # Use of @login_required decorator
def get_modules_page(roomid):
    '''
    Show the modules for a given room
    :return:
    '''
    #modules = get_modules_for_room(roomid)
    room = get_room(roomid)
    all_rooms = get_rooms(get_current_house()["houseid"])
    return render_template('modules.html', room=room, rooms=all_rooms, ngapp="ahps_web", ngcontroller="modulesController")


@app.route('/modules/<roomid>', methods=['GET'])
@login_required                                 # Use of @login_required decorator
def get_modules(roomid):
    modules = get_modules_for_room(roomid)
    return jsonify({"modules": modules})


@app.route('/modules/<moduleid>/state', methods=['PUT'])
@login_required                                 # Use of @login_required decorator
def set_module_state(moduleid):
    """
    Change state of module to on or off
    :param roomid:
    :return:
    """
    args = json.loads(request.data.decode())["data"]
    if args["state"] == "on":
        if device_on(moduleid):
            # Return success
            pass
        else:
            # Return error
            pass
    elif args["state"] == "off":
        if device_off(moduleid):
            # Return success
            pass
        else:
            # Return error
            pass
    else:
        # Return an error
        pass
    return ""


# @app.route('/modules/<roomid>', methods=['POST'])
# @login_required                                 # Use of @login_required decorator
# def post_modules(roomid):
#     '''
#     Show the modules page for a given room
#     :return:
#     '''
#     # Possible actions: Save, edit programs, remove, on, off
#     moduleid = request.form["moduleid"]
#     button = request.form["button"]
#
#     if button == 'save':
#         update_module_type(moduleid, request.form["module_type"])
#         update_module_name(moduleid, request.form["module-name"])
#         if request.form['module_type'] == 'house':
#             update_module_hdc(moduleid, request.form["house_code"], "")
#         else:
#             update_module_hdc(moduleid, request.form["house_code"], request.form["device_code"])
#         if request.form['module_type'] == 'lamp':
#             update_module_dim_amount(moduleid, request.form['dim_amount'])
#         flash('Module record saved')
#
#     elif button == 'editprograms':
#         return redirect(url_for("module_programs", moduleid=moduleid))
#
#     elif button == 'remove':
#         module = get_module(moduleid)
#         # This is a cascading delete
#         delete_module(moduleid)
#         flash("The \"{0}\" module was removed".format(module['name']))
#
#     elif button == 'on':
#         if request.form['module_type'] != 'house':
#             # Appliance or lamp
#             if device_on(moduleid):
#                 flash("{0} turned on".format(request.form['module_type']))
#         else:
#             # All lights on for house code
#             if all_lights_on(moduleid):
#                 flash("All lights turned on")
#
#     elif button == 'off':
#         if request.form['module_type'] != 'house':
#             if device_off(moduleid):
#                 flash("{0} turned off".format(request.form['module_type']))
#         else:
#             # All lights off for house code
#             if all_lights_off(moduleid):
#                 flash("All lights turned off")
#
#     else:
#         return "Unrecognized button action"
#
#     modules = get_modules_for_room(roomid)
#     room = get_room(roomid)
#     all_rooms = get_rooms(get_current_house()["houseid"])
#     return render_template('modules.html', room=room, modules=modules, rooms=all_rooms)


@app.route('/module/<moduleid>', methods=['PUT'])
@login_required                                 # Use of @login_required decorator
def put_module(moduleid):
    """
    Update a module
    :param moduleid:
    :return:
    """
    args = json.loads(request.data.decode())["data"]

    update_module_type(moduleid, args["module_type"])
    update_module_name(moduleid, args["name"])
    if args['module_type'] == 'house':
        update_module_hdc(moduleid, args["house_code"], "")
    else:
        update_module_hdc(moduleid, args["house_code"], args["device_code"])
    if args['module_type'] == 'lamp':
        update_module_dim_amount(moduleid, args['dim_amount'])

    return ""


@app.route('/module/<moduleid>', methods=['DELETE'])
@login_required                                 # Use of @login_required decorator
def delete_room_module(moduleid):
    # This is a cascading delete
    delete_module(moduleid)
    return ""


@app.route('/modules/edit_module', methods=['POST'])
@login_required                                 # Use of @login_required decorator
def edit_module():
    return 'Edit module'


@app.route('/room/<roomid>/new_appliance_module', methods=['GET'])
@login_required                                 # Use of @login_required decorator
def add_appliance_module(roomid):
    room = get_room(roomid)
    return render_template('new_appliance_module.html', roomid=room['roomid'], room_name=room['name'],
                           ngapp="ahps_web", ngcontroller="addApplianceController")


@app.route('/room/<roomid>/new_lamp_module', methods=['GET'])
@login_required                                 # Use of @login_required decorator
def add_lamp_module(roomid):
    room = get_room(roomid)
    return render_template('new_lamp_module.html', roomid=room['roomid'], room_name=room['name'],
                           ngapp="ahps_web", ngcontroller="addLampController")


@app.route('/room/<roomid>', methods=["POST"])
@login_required
def create_module(roomid):
    """
    Add a new module to a room.
    :return:
    """
    args = json.loads(request.data.decode())["data"]

    if args["module_type"] == "appliance":
        insert_module(roomid, args["module_type"], args['name'],
                      args['house_code'], args['device_code'])
    elif args["module_type"] == "lamp":
        insert_module(roomid, args["module_type"], args['name'],
                      args['house_code'], args['device_code'], args["dim_amount"])
    else:
        # House?
        pass

    return ""


@app.route('/modules/page', methods=["GET"])
@login_required                                 # Use of @login_required decorator
def all_house_codes():
    house = get_current_house()
    return render_template("all_house_codes.html", house=house, ngapp="ahps_web", ngcontroller="allModulesController")


@app.route('/modules', methods=["GET"])
@login_required                                 # Use of @login_required decorator
def get_modules_for_current_house():
    house = get_current_house()
    mods = get_modules_for_house(house["houseid"])
    return jsonify({"modules": mods})


@app.route('/modules/selected', methods=["PUT"])
@login_required                                 # Use of @login_required decorator
def save_selected():
    """
    Form actions on list of all house codes. The user
    can alter the list of selected modules, turn selected
    modules on or turn them off.
    :return:
    """
    args = json.loads(request.data.decode())["data"]
    # Handle save, on, off cases
    for module in args:
        moduleid = module["moduleid"]
        update_module_selected(module["moduleid"], module["selected"])

    return ""


@app.route('/modules/on', methods=["PUT"])
@login_required                                 # Use of @login_required decorator
def lights_on():
    """
    Turn on all of the selected modules in the current house
    """
    house = get_current_house()
    mods = get_modules_for_house(house["houseid"])
    for module in mods:
        if module["selected"]:
            device_on(module["moduleid"])
    return ""


@app.route('/modules/off', methods=["PUT"])
@login_required                                 # Use of @login_required decorator
def lights_off():
    """
    Turn off all of the selected modules in the current house
    """
    house = get_current_house()
    mods = get_modules_for_house(house["houseid"])
    for module in mods:
        if module["selected"]:
            device_off(module["moduleid"])
    return ""


@app.route('/rooms/<roomid>/module/<moduleid>', methods=['POST'])
@login_required                                 # Use of @login_required decorator
def move_module(roomid, moduleid):
    """
    Move a module to another room.
    This URL probably belongs in the home view, but since it is used
    on the modules page we'll leave it here.
    :param roomid:
    :param moduleid:
    :return:
    """
    # Implement move to new room
    move_module_room(moduleid, roomid)

    return ""
